core/lib/zksync_core/src/l1_gas_price/gas_adjuster/validium_gas_adjuster.rs

```diff
@@ -0,0 +1,61 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use zksync_config::GasAdjusterConfig;
+use zksync_eth_client::{Error, EthInterface};
+
+use super::{GasAdjuster, GasStatistics, L1TxParamsProvider};
+
+#[derive(Debug)]
+pub struct ValidiumGasAdjuster {
+    pub(super) statistics: GasStatistics,
+    pub(super) config: GasAdjusterConfig,
+    eth_client: Arc<dyn EthInterface>,
+}
+
+#[async_trait]
+impl GasAdjuster for ValidiumGasAdjuster {
+    fn config(&self) -> &GasAdjusterConfig {
+        &self.config
+    }
+
+    fn eth_client(&self) -> &dyn EthInterface {
+        &self.eth_client
+    }
+
+    fn statistics(&self) -> &GasStatistics {
+        &self.statistics
+    }
+
+    fn into_l1_tx_params_provider(self: Arc<Self>) -> Arc<dyn L1TxParamsProvider> {
+        self
+    }
+
+    fn estimate_effective_pubdata_price(&self) -> u64 {
+        0
+    }
```

Comment:
This seems to be the only difference in behavior for now, with a lot of other interfaces being exposes just for the sake of avoiding code duplication.
However, we can also avoid code duplication by abstracting over the things that are different. In this case, the pubdata pricing. So, we can define a trait for that specifically:

```rust
pub trait PubdataPricing {
    /// Returns the amount of L1 gas to publish a single byte of pub data.
    fn pubdata_byte_gas(&self) -> u64;
}
```

Then we can have `ValidiumPubdataPricing` and `RollupPubdataPricing` implementing this trait.
`GasAdjuster` will no longer have to be a trait as a whole, but instead it'll hold `Box<dyn PubdataPricing>` inside, which will be called inside of `estimate_effective_pubdata_price`.

Reply:
I agree with this, it is a much simpler approach. We came to this solution from what was discussed here https://github.com/matter-labs/zksync-era/pull/1017#discussion_r1481485699 with @RomanBrodetski. Let us know what you prefer and we'll go in that direction.

Reply:
Yeah, I've seen this discussion, and I think that @RomanBrodetski in fact had something like this in mind, he just described a very rough draft of the idea (e.g. even in the current implementation, most of functionality remains in GasAdjuster, so it doesn't need to be abstracted itself, only its parts).

Let's switch to having a trait for pubdata pricing only.